hydra-cluster/test/Test/HydraExplorerSpec.hs

```diff
@@ -0,0 +1,239 @@
+              bobHeadId <- withHydraNode hydraTracer bobChainConfig tmpDir 2 bobSk [] [2] $ \hydraNode -> do
+                seedFromFaucet_ cardanoNode bobCardanoVk 100_000_000 (contramap FromFaucet tracer)
+
+                send hydraNode $ input "Init" []
+
+                bobHeadId <- waitMatch 5 hydraNode $ \v -> do
+                  guard $ v ^? key "tag" == Just "HeadIsInitializing"
+                  v ^? key "headId" . _String
+
+                headExplorerSees explorer "HeadInitTx" bobHeadId
+
+                send hydraNode $ input "Abort" []
+
+                headExplorerSees explorer "HeadAbortTx" bobHeadId
+
+                pure bobHeadId
+
+              manager <- HTTPClient.newTlsManager
+              let url = "http://127.0.0.1:9090/heads"
+              request <-
+                HTTPClient.parseRequest url <&> \request ->
+                  request
+                    { HTTPClient.method = "GET"
+                    , HTTPClient.requestHeaders = [("Accept", "application/json")]
+                    }
+              response <- HTTPClient.httpLbs request manager
+              HTTPClient.responseStatus response `shouldBe` status200
+              allHeads <- unsafeDecodeJson . toStrict $ HTTPClient.responseBody response
+              length allHeads `shouldBe` 2
+              let [ HeadState{headId = aliceIdBS, status = aliceStatus}
+                    , HeadState{headId = bobIdBS, status = bobStatus}
+                    ] = allHeads
+              encode aliceHeadId `shouldBe` encode aliceIdBS
+              aliceStatus `shouldBe` Initializing
+              encode bobHeadId `shouldBe` encode bobIdBS
+              bobStatus `shouldBe` Aborted
+
+headExplorerSees :: HasCallStack => HydraExplorerHandle -> Value -> Text -> IO ()
+headExplorerSees explorer txType headId =
+  awaitMatch explorer 5 $ \v -> do
+    guard $ v ^? key "message" . key "tag" == Just txType
+    let actualId = v ^? key "message" . key "headId" . _String
+    guard $ actualId == Just headId
+
+awaitMatch :: HasCallStack => HydraExplorerHandle -> DiffTime -> (Aeson.Value -> Maybe a) -> IO a
+awaitMatch hydraExplorerHandle delay f = do
+  seenMsgs <- newTVarIO []
+  timeout delay (go seenMsgs) >>= \case
+    Just x -> pure x
+    Nothing -> do
+      msgs <- readTVarIO seenMsgs
+      failure $
+        toString $
+          unlines
+            [ "awaitMatch did not match a message within " <> show delay
+            , padRight ' ' 20 "  seen messages:"
+                <> unlines (align 20 (decodeUtf8 . Aeson.encode <$> msgs))
+            ]
+ where
+  go seenMsgs = do
+    msg <- awaitNext hydraExplorerHandle
+    atomically (modifyTVar' seenMsgs (msg :))
+    maybe (go seenMsgs) pure (f msg)
+
+  align _ [] = []
+  align n (h : q) = h : fmap (T.replicate n " " <>) q
+
+newtype HydraExplorerHandle = HydraExplorerHandle {awaitNext :: IO Value}
+
+data HydraExplorerLog
```

Comment:
There might be a better place for these types.

Reply:
It's a bit hard to move them since they depend on hydra-cluster...Let's leave them as is.